### PR TITLE
Implement mute/unmute room and add offer to join

### PR DESCRIPTION
**Pull request by golgetahir**
Hi team, thanks for the implementation helped me greatly. Though in the audio bridge plugin, there are some events not there so I added mute/unmute with event callback to peers and generate_offer flag to join room. I am sharing this PR with you hoping that it will be merged or help in any way. Thanks a lot for great work!

### CODE REVIEW

File: src/plugins/audiobridge-plugin.js [hunk 1]
```diff
@@ -52,6 +54,7 @@ const PLUGIN_EVENT = {
   ALLOWED: 'audiobridge_allowed',
   SUCCESS: 'audiobridge_success',
   ERROR: 'audiobridge_error',
+  ROOM_MUTED: 'audiobridge_room_muted'
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:57]: As a convention, I'd prefer having `SUCCESS` and `ERROR` as last events, so could you please move `ROOM_MUTED` two positions up ?

File: src/plugins/audiobridge-plugin.js [hunk 2]
```diff
@@ -296,6 +299,13 @@ class AudioBridgeHandle extends Handle {
             janode_event.event = PLUGIN_EVENT.PEER_LEAVING;
             break;
           }
+          /* Room muted event */
+          if (typeof message_data.muted !== 'undefined') {
+            janode_event.data.room = message_data.room;
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:304]: This is not necessary, the message handler already fills the `room` property before parsing the events

File: src/plugins/audiobridge-plugin.js [hunk 3]
```diff
@@ -787,6 +798,50 @@ class AudioBridgeHandle extends Handle {
     throw (error);
   }
 
+  /**
+   *
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:802]: Missing description of the function

File: src/plugins/audiobridge-plugin.js [hunk 4]
```diff
@@ -787,6 +798,50 @@ class AudioBridgeHandle extends Handle {
     throw (error);
   }
 
+  /**
+   * 
+   * @param {object} params
+   * @param {number|string} params.room - The involved room
+   * @param {string} [params.secret] - The optional secret needed to manage the room
+   * @returns {} - Returns event as success or error
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:806]: This should return something like

`Promise<module:audiobridge-plugin~AUDIOBRIDGE_EVENT_MUTE_ROOM_RESPONSE>`

with a simple definition of `AUDIOBRIDGE_EVENT_MUTE_ROOM_RESPONSE` as
```
/**
 * The response event for audiobridge mute request.
 *
 * @typedef {object} AUDIOBRIDGE_EVENT_MUTE_ROOM_RESPONSE
 * @property {number|string} room - The involved room
 */
```

File: src/plugins/audiobridge-plugin.js [hunk 5]
```diff
@@ -350,7 +360,7 @@ class AudioBridgeHandle extends Handle {
    * @param {string} [params.group] - The group to assign to this participant
    * @returns {Promise<module:audiobridge-plugin~AUDIOBRIDGE_EVENT_JOINED>}
    */
-  async join({ room, feed, display, muted, pin, token, quality, volume, record, filename, rtp_participant, group }) {
+  async join({ room, feed, display, muted, pin, token, quality, volume, record, filename, rtp_participant, group, generate_offer }) {
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:363]: Missing docs for `generate_offer` parameter

File: src/plugins/audiobridge-plugin.js [hunk 6]
```diff
@@ -787,6 +798,50 @@ class AudioBridgeHandle extends Handle {
...
+    const response = await this.message(body);
+    const { event, data: evtdata } = response._janode || {};
+    if (event === PLUGIN_EVENT.SUCCESS)
+      return evtdata;
+    const error = new Error(`unexpected response to ${body.request} request`);
+    throw (error);
+  }
+
+   /**
+   *
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:824]: Missing description

File: src/plugins/audiobridge-plugin.js [hunk 7]
```diff
@@ -787,6 +798,50 @@ class AudioBridgeHandle extends Handle {
...
+    const error = new Error(`unexpected response to ${body.request} request`);
+    throw (error);
+  }
+
+   /**
+   * 
+   * @param {object} params
+   * @param {number|string} params.room - The involved room
+   * @param {string} [params.secret] - The optional secret needed to manage the room
+   * @returns {} - Returns event as success or error
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:828]: Similarly to the mute function, this should return a documented

`Promise<module:audiobridge-plugin~AUDIOBRIDGE_EVENT_UNMUTE_ROOM_RESPONSE>`

File: src/plugins/audiobridge-plugin.js [hunk 8]
```diff
@@ -959,7 +1014,9 @@ class AudioBridgeHandle extends Handle {
  * @property {string} EVENT.AUDIOBRIDGE_TALKING {@link module:audiobridge-plugin~AUDIOBRIDGE_TALKING}
  * @property {string} EVENT.AUDIOBRIDGE_PEER_TALKING {@link module:audiobridge-plugin~AUDIOBRIDGE_PEER_TALKING}
  * @property {string} EVENT.AUDIOBRIDGE_ERROR {@link module:audiobridge-plugin~AUDIOBRIDGE_ERROR}
+ * @property {string} EVENT.AUDIOBRIDGE_ROOM_MUTED {@link module:audiobridge-plugin~AUDIOBRIDGE_ROOM_MUTED}
  */
+
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:1019]: Please remove this empty line

File: src/plugins/audiobridge-plugin.js [hunk 9]
```diff
@@ -959,7 +1014,9 @@ class AudioBridgeHandle extends Handle {
  * @property {string} EVENT.AUDIOBRIDGE_TALKING {@link module:audiobridge-plugin~AUDIOBRIDGE_TALKING}
  * @property {string} EVENT.AUDIOBRIDGE_PEER_TALKING {@link module:audiobridge-plugin~AUDIOBRIDGE_PEER_TALKING}
  * @property {string} EVENT.AUDIOBRIDGE_ERROR {@link module:audiobridge-plugin~AUDIOBRIDGE_ERROR}
+ * @property {string} EVENT.AUDIOBRIDGE_ROOM_MUTED {@link module:audiobridge-plugin~AUDIOBRIDGE_ROOM_MUTED}
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:1017]: Please keep ERROR as the last event, just move muted one position up

File: src/plugins/audiobridge-plugin.js [hunk 10]
```diff
@@ -1058,5 +1115,15 @@ export default {
      * @type {Error}
      */
     AUDIOBRIDGE_ERROR: PLUGIN_EVENT.ERROR,
+
+    /**
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:1119]: Please keep ERROR in the last position

**Review by atoppi — CHANGES_REQUESTED** [review]
Thank for the effort @golgetahir.
I have left some comments that should be addressed before merging the PR.

**Comment by golgetahir** [comment]
@atoppi Thanks for the review, I think I have done it accordingly now.

**Comment by atoppi** [comment]
Thanks looks good now!
I'll merge it once I have the chance to check that everything works as expected.

**Review by atoppi — APPROVED** [review]

File: src/plugins/audiobridge-plugin.js [hunk 11]
```diff
@@ -50,8 +52,9 @@ const PLUGIN_EVENT = {
   RTP_FWD: 'audiobridge_rtp_fwd',
   FWD_LIST: 'audiobridge_rtp_list',
   ALLOWED: 'audiobridge_allowed',
+  ROOM_MUTED: 'audiobridge_room_muted',
   SUCCESS: 'audiobridge_success',
-  ERROR: 'audiobridge_error',
+  ERROR: 'audiobridge_error'
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:57]: missing trailing comma

File: src/plugins/audiobridge-plugin.js [hunk 12]
```diff
@@ -26,6 +26,8 @@ const REQUEST_ALLOW = 'allowed';
 const REQUEST_RTP_FWD_START = 'rtp_forward';
 const REQUEST_RTP_FWD_STOP = 'stop_rtp_forward';
 const REQUEST_RTP_FWD_LIST = 'listforwarders';
+const REQUEST_MUTE_ROOM = 'mute_room'
+const REQUEST_UNMUTE_ROOM = 'unmute_room'
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:30]: missing semicolons for both lines

File: src/plugins/audiobridge-plugin.js [hunk 13]
```diff
@@ -787,6 +798,52 @@ class AudioBridgeHandle extends Handle {
...
+    const body = {
+      request: REQUEST_MUTE_ROOM,
+      room,
+    };
+    if (typeof secret === 'string') body.secret = secret;
+
+    const response = await this.message(body);
+    const { event, data: evtdata } = response._janode || {};
+    if (event === PLUGIN_EVENT.SUCCESS)
+      return evtdata;
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:819]: according to janus docs, room is not part of the success response.
Could you please copy the `room` id from the request body? e.g.

```
if (event === PLUGIN_EVENT.SUCCESS) {
      evtdata.room = body.room;
      return evtdata;
}
```

File: src/plugins/audiobridge-plugin.js [hunk 14]
```diff
@@ -787,6 +798,52 @@ class AudioBridgeHandle extends Handle {
...
+      const body = {
+        request: REQUEST_UNMUTE_ROOM,
+        room,
+      };
+      if (typeof secret === 'string') body.secret = secret;
+  
+      const response = await this.message(body);
+      const { event, data: evtdata } = response._janode || {};
+      if (event === PLUGIN_EVENT.SUCCESS)
+        return evtdata;
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:842]: same as mute room, please add the missing room attribute

**Review by atoppi — CHANGES_REQUESTED** [review]
Everything seems to work as intended.
I've left some other notes.

The sample code under `examples/audiobridge` is missing.
I will add it once the PR has been merged.

**Comment by golgetahir** [comment]
Thanks for review, changes are done.

File: src/plugins/audiobridge-plugin.js [hunk 15]
```diff
@@ -787,6 +798,54 @@ class AudioBridgeHandle extends Handle {
...
+  async muteRoom({ room, secret }) {
+    const body = {
+      request: REQUEST_MUTE_ROOM,
+      room,
+    };
+    if (typeof secret === 'string') body.secret = secret;
+
+    const response = await this.message(body);
+    const { event, data: evtdata } = response._janode || {};
+    if (event === PLUGIN_EVENT.SUCCESS)
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:818]: Missing curly brackets for the if block

File: src/plugins/audiobridge-plugin.js [hunk 16]
```diff
@@ -787,6 +798,54 @@ class AudioBridgeHandle extends Handle {
...
+    async unmuteRoom({ room, secret }) {
+      const body = {
+        request: REQUEST_UNMUTE_ROOM,
+        room,
+      };
+      if (typeof secret === 'string') body.secret = secret;
+  
+      const response = await this.message(body);
+      const { event, data: evtdata } = response._janode || {};
+      if (event === PLUGIN_EVENT.SUCCESS)
```
Comment (atoppi) [inline src/plugins/audiobridge-plugin.js:842]: Missing curly brackets for the if block

**Review by atoppi — COMMENTED** [review]
Thanks, it seems you missed some curly brackets

**Review by atoppi — APPROVED** [review]

**Comment by atoppi** [comment]
fyi in https://github.com/meetecho/janode/commit/903cd75b285ef160e646bba1138ddbdc5c611437 I have also added the sample code for this feature and corrected some indentation issues.
Thanks again!

**Pull request merged**